src/makerom/entry.s

```diff
@@ -1,34 +1,45 @@
 #include "ultra64/asm.h"
+#include "ultra64/regdef.h"
 #include "boot.h"
 
-.set noreorder
+.text
 
-.section .text
-
-.balign 16
-
-LEAF(entrypoint)
-    // Clear boot segment .bss
-    la      $t0, _bootSegmentBssStart
-#ifndef AVOID_UB
-    // UB: li only loads the lower 16 bits of _bootSegmentBssSize when it may be larger than this,
-    // so not all of bss may be cleared if it is too large
-    li      $t1, _bootSegmentBssSize
+#if defined(__sgi) && !defined(AVOID_UB)
+/* IDO assembler workaround: The makerom tool in the N64 SDK was given the bss segment size as a const
+ * literal, and since this literal was < 0x10000 it was loaded in one instruction. We don't have access
+ * to the bss segment size until we link everything so we cannot do the same thing. We can load the
+ * lower 16 bits of the bss size; however IDO emits two R_MIPS_LO16 relocations at this instruction,
+ * so we must use a special symbol that is the bss size divided by two so that the linker updates the
+ * value twice to the correct bss size.
+ * When AVOID_UB is enabled, don't do this and instead load the full symbol value. */
+#define LOAD_BSS_SIZE(reg) addiu reg, zero, %lo(_bootSegmentBssSize_2)
 #else
-    la      $t1, _bootSegmentBssSize
+#define LOAD_BSS_SIZE(reg) la reg, _bootSegmentBssSize
 #endif
+
+LEAF(entrypoint)
+    /* Clear boot segment .bss */
+    la      t0, _bootSegmentBssStart
+    LOAD_BSS_SIZE(t1)
 .clear_bss:
-    addi    $t1, $t1, -8
-    sw      $zero, ($t0)
-    sw      $zero, 4($t0)
-    bnez    $t1, .clear_bss
-     addi   $t0, $t0, 8
-    // Set up stack and enter program code
-    lui     $t2, %hi(bootproc)
-    lui     $sp, %hi(sBootThreadStack + BOOT_STACK_SIZE)
-    addiu   $t2, %lo(bootproc)
-    jr      $t2
-     addiu  $sp, %lo(sBootThreadStack + BOOT_STACK_SIZE)
+    sw      zero, 0(t0)
+    sw      zero, 4(t0)
+    addi    t0, t0, 8
+    addi    t1, t1, -8
+    bnez    t1, .clear_bss
+
+    /* Set up stack and enter program code */
+    la      sp, sBootThreadStack + BOOT_STACK_SIZE
+    la      t2, bootproc
+    jr      t2
 END(entrypoint)
 
+#ifdef __GNUC__
+/* Pad to a total size of 0x60 */
 .fill 0x60 - (. - entrypoint)
+#else
+/* IDO can't take absolute differences of symbols.. */
+.repeat (0x60 - 0x34)
+    .byte 0
+.endr
+#endif
```

Comment:
this is maybe not relevant to this PR, but I noticed on iQue that the entry code is shorter and the address of boot is 0x80000450, not 0x80000460. This makes me think we should manually add a bunch of NOPs here and/or let the linker round up the size to 0x10, and stop hardcoding the address of boot in the spec. That would clean up the ugliness here too

Reply:
I don't think it'll be particularly cleaner either way, and the old gas version for egcs might demand its own solution here as well, so I don't have enough context to really consider what's better here rn. I plan to work on matching makerom next so this is probably better left for that.